src/Lecture4.hs

```diff
@@ -135,7 +140,33 @@ errors. We will simply return an optional result here.
 -}
 
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = readRow . splitCSVLine
+   where readRow :: [String] -> Maybe Row
+         readRow cols
+            | length cols /= 3 = Nothing
+            | otherwise = Row <$> readProduct c1 <*> readTradeType c2 <*> readInt c3
+               where c1 = head cols
+                     c2 = cols !! 1
+                     c3 = last cols
+         readProduct :: String -> Maybe String
+         readProduct p
+            | dropSpaces p == "" = Nothing
+            | otherwise = Just p
+         readTradeType :: String -> Maybe TradeType
+         readTradeType trade = case dropSpaces trade of
+            "Sell" -> Just Sell
+            "Buy"  -> Just Buy
+            _      -> Nothing
+         readInt :: String -> Maybe Int
+         readInt int = readMaybe int >>= positiveOrNothing
+         positiveOrNothing :: Int -> Maybe Int
+         positiveOrNothing x = if x < 0 then Nothing else Just x
```

Comment:
You have an excellent monadic (and even Applicative!) implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow str = do
    [c1, c2, c3] <- Just $ splitCSVLine str
    guard $ c1 /= ""
    trade <- readMaybe c2
    cost <- readMaybe c3
    guard $ cost >= 0
    pure $ Row c1 trade cost
```

Reply:
wow, that looks nice! I will check the post! thanks.